From: beyondmagic <koetemagie@gmail.com>
Date: Tue, 04 Nov 2021
Subject: [PATCH] Patch to open a new terminal with the same path.
[CUSTOM] specify command.
diff --git a/config.def.h b/config.def.h
index 6f05dce..4d3a822 100644
--- a/config.def.h
+++ b/config.def.h
@@ -199,6 +199,8 @@ static Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_Y,           selpaste,       {.i =  0} },
 	{ ShiftMask,            XK_Insert,      selpaste,       {.i =  0} },
 	{ TERMMOD,              XK_Num_Lock,    numlock,        {.i =  0} },
+  { ShiftMask,            XK_BackSpace,   newterm,        {.v = "st_float"} },
+  { ShiftMask,            XK_Return,      newterm,        {.v ="st"} },
 };
 
 /*
diff --git a/st.c b/st.c
index a9338e1..a9a9f9b 100644
--- a/st.c
+++ b/st.c
@@ -153,6 +153,7 @@ typedef struct {
 } STREscape;
 
 static void execsh(char *, char **);
+static char *getcwd_by_pid(pid_t pid);
 static void stty(char **);
 static void sigchld(int);
 static void ttywriteraw(const char *, size_t);
@@ -1061,6 +1062,26 @@ tswapscreen(void)
 	tfulldirt();
 }
 
+void
+newterm(const Arg *arg)
+{
+	switch (fork()) {
+	case -1:
+		die("fork failed: %s\n", strerror(errno));
+		break;
+	case 0:
+		chdir(getcwd_by_pid(pid));
+		execlp(arg->v, arg->v, NULL);
+		break;
+	}
+}
+
+static char *getcwd_by_pid(pid_t pid) {
+	char buf[32];
+	snprintf(buf, sizeof buf, "/proc/%d/cwd", pid);
+	return realpath(buf, NULL);
+}
+
 void
 tscrolldown(int orig, int n)
 {
diff --git a/st.h b/st.h
index fa2eddf..b13399b 100644
--- a/st.h
+++ b/st.h
@@ -81,6 +81,7 @@ void die(const char *, ...);
 void redraw(void);
 void draw(void);
 
+void newterm(const Arg *);
 void printscreen(const Arg *);
 void printsel(const Arg *);
 void sendbreak(const Arg *);
